### ISSUE TITLE
Add wrapper for `GetPrivateProfileInt` and `WritePrivateProfileString`

### ISSUE THREAD

**Issue by DGSZink** [mkxp-z/mkxp-z#12]
The script that I'm trying to use is "Master Volume Control for RGSS3 by むっくRu (Rutan)" that allows the user to change sound volumes separately in-game, then saves the changes in the game ini file and it just loads that on boot.

The only issue is that it gets an error with mkxp-z because GetPrivateProfileInt and WritePrivateProfileString are kernel32 functions and they doesn't work on things like Linux or Mac.

Can support for those be added here or make the script remember the changes without it so it works multiplatform?

**Comment by ghost** [comment]
I'd just rewrite it to include settings within mkxp.json instead: https://github.com/mkxp-z/mkxp-z/wiki/Extensions-(New)#cfg-overriding-reading-from-and-saving-to-the-json

Maybe try that? Unless there's a particular reason why you just really want to use INI

**Comment by Splendide-Imaginarius** [comment]
If, for whatever reason, you don't want to rewrite the volume control script, you could try implementing the relevant Windows API functions in a preload script. See #6 for some possible inspiration. Once that PR gets through review, the `win32_wrap.rb` script seems like a good place to put the functions you're missing.

**Comment by Splendide-Imaginarius** [comment]
@DGSZink Can you point me to a game that uses those API functions? I might be interested in adding support for them.

**Comment by Splendide-Imaginarius** [comment]
Well, in a stroke of bad (good?) fortune, I bought a game yesterday, tried to play it on Linux, noticed that the audio was muted, inspected logs, and sure enough, it seems to use exactly the script you mention. So I guess now this is on my to-do list.

I think the most reasonable approach to getting that script working is a preload script that converts the Windows INI API calls to the JSON API that @strumajen suggested. It won't produce the same underlying filesystem behavior as what Windows does, but the game won't see any difference.

Anyway I probably won't get to this for some time; too many other things in the queue.

**Comment by Splendide-Imaginarius** [comment]
I have this partially working now, at least enough that the audio is no longer artificially muted. Unfortunately the game I'm testing with still errors even with the fix, so my motivation (and ability to properly test) is diminished, but I'll see if I can finish up the implementation of this and send in a PR for you to try. No promises on timeline though.

**Comment by maxchang3** [comment]
I encountered a similar problem where my game and your game used the same script. Following the suggestions above, I added [scripts/preload/win32_wrap.rb](https://github.com/mkxp-z/mkxp-z/blob/52c932b9a06b484aa7cb708c254fe3c9a08e1f1c/scripts/preload/win32_wrap.rb#L336) as a preload script, and my game could start normally! 

However, there was no sound, even though I could modify the volume in the game, but it wasn't saved. Then, I noticed in the logs that there were attempts to call the two functions mentioned in this issue. So, I started working on a solution.

Firstly, I added two functions for handling INI file operations (since I'm not familiar with Ruby, these were generated by Copilot):

```ruby
def read_ini(section, parameter, default, file_location)
  current_section = nil
  File.open(file_location, "r") do |file|
    file.each_line do |line|
      line.strip!
      if line.start_with?("[") && line.end_with?("]")
        current_section = line[1..-2]
      elsif current_section == section
        key, value = line.split("=").map(&:strip)
        return value.to_i if key == parameter
      end
    end
  end
  default
end

def write_ini(section, parameter, value, file_location)
  data = {}
  if File.exist?(file_location)
    File.open(file_location, 'r') do |file|
      current_section = nil
      file.each_line do |line|
        line.strip!
        if line.start_with?('[') && line.end_with?(']')
          current_section = line[1..-2]
          data[current_section] = {}
        elsif current_section
          key, val = line.split('=').map(&:strip)
          data[current_section][key] = val
        end
      end
    end
  end

  data[section] ||= {}
  data[section][parameter] = value.to_s unless value.nil? || value.to_s.strip.empty?

  File.open(file_location, 'w') do |file|
    data.each do |section, parameters|
      file.puts "[#{section}]"
      parameters.each do |key, value|
        file.puts "#{key}=#{value}" unless value.nil? || value.to_s.strip.empty?
      end
      file.puts
    end
  end
end
```

Then, I added the following two classes within the module Win32API_Impl:

```diff
+ module Kernel32  
+   class GetPrivateProfileInt  
+     def call(args)  
+       section, parameter, default, file_location = args  
+       read_ini(section, parameter, default, file_location)  
+     end  
+   ed

+     
+   class WritePrivateProfileString  
+     def call(args)  
+       section, parameter, value, file_location = args  
+       write_ini(section, parameter, value, file_location)  
+     end  
+   end
+ end
```

After these changes, my game had volume in the startup interface and could modify and save values in Game.ini!

I hope this helps!

**Comment by WaywardHeart** [comment]
That's a lot simpler than what I ended up using for any games that need it (the [inifile gem](https://github.com/TwP/inifile), with a small change to remove invalid characters if needed.). It probably doesn't cover cases like multi-line values, though, although I doubt those happen often if at all.

Joiplay actually implements these and a bunch of other functions in c++, if anyone's interested in bringing that over to restore our built-in Win32API support.

**Comment by maxchang3** [comment]
Yes, I implemented it roughly just to test if it works, focusing  on this specific issue. This is definitely not the final solution.   Bring the implementation from Joiplay seems like a viable option.

**Comment by Splendide-Imaginarius** [comment]
I haven't looked at Joiplay's implementation, but in principle I have no problem with merging improvements from them as long as they're reasonably high-quality. Implementing cross-platform Win32API in C++ seems like a major endeavor (and a pretty big scope expansion), but if they've already done the work and we're not expected to expand it ourselves, it's not necessarily a dealbreaker.
